reconcile/external_resources/secrets_sync.py

```diff
@@ -99,54 +118,95 @@ def _init_ocmap(self, specs: Iterable[ExternalResourceSpec]) -> OCMap:
             integration=QONTRACT_INTEGRATION,
         )
 
-    @retry()
-    def _write_secrets_to_vault(self, spec: ExternalResourceSpec) -> None:
-        if spec.secret:
-            secret_path = f"{self.vault_path}/{QONTRACT_INTEGRATION}/{spec.cluster_name}/{spec.namespace_name}/{spec.output_resource_name}"
-            stringified_secret = {k: str(v) for k, v in spec.secret.items()}
-            desired_secret = {"path": secret_path, "data": stringified_secret}
-            self.vault_client.write(desired_secret, decode_base64=False)
+    def _copy_spec(self, spec: ExternalResourceSpec) -> ExternalResourceSpec:
+        copy = dataclasses.replace(spec)
+        copy.secret = dict(spec.secret)
+        return copy
 
-    def sync_secrets(self, specs: Iterable[ExternalResourceSpec]) -> None:
+    def sync_secrets(
+        self, specs: Iterable[ExternalResourceSpec]
+    ) -> list[ExternalResourceSpec]:
         self._populate_secret_data(specs)
-        ri = ResourceInventory()
-        self._initialize_ri(ri, specs)
-        ocmap = self._init_ocmap(specs)
-        for item in ri:
-            self.reconcile_data(item, ri, ocmap)
+
+        # Updated_at attribute must be removed before syncing the secret
+        # but is needed afterwards. All specs are copied to preserve the originals.
+        to_sync_specs = [
+            self._copy_spec(spec) for spec in self._specs_with_secret(specs)
+        ]
+
+        for spec in to_sync_specs:
+            self._annotate(spec)
+            self._del_secret_updated_at(spec)
```

Comment:
this is needed because we embed update time into secret data, would it make sense to add a new root field to spec so we don't need to do add and remove?

Reply:
I thought optimistic locking only needed in vault to cluster, in code it's reused for both reconciler, but shouldn't be any issues